Add serialization support for common::Filter

duanmeng opened the pull request:
Details in https://github.com/facebookincubator/velox/issues/4614

mbasmanova reviewed (commented):
@Yuhta Jimmy, would you help review this PR?

Yuhta reviewed (commented):
Try remove the unused constructors and reuse default copy/move ctor as much as possible

duanmeng commented:
> Try remove the unused constructors and reuse default copy/move ctor as much as possible

@Yuhta Hi Jimmy,  I have addressed all the comments according to your guidance, refactored the `testingEquals` function, removed the redundant copy/move constructors for `BigintValuesUsingHashTable` and `BigintValuesUsingBitmask`, and used their default constructors in `NegatedBigintValuesUsingHashTable` and `NegatedBigintValuesUsingBitmask`. And I've explained why to use FilterPtr as the `create` function's return type, below the comment you left. Could you please help to take a look?

Yuhta commented on velox/type/Filter.h:
```diff
@@ -261,6 +310,10 @@ class AlwaysFalse final : public Filter {
 
+  folly::dynamic serialize() const override;
+
+  static FilterPtr create(const folly::dynamic& /*obj*/);
```
We should change `deserialize` here to return `unique_ptr` then.  `unique_ptr` can be moved into `shared_ptr` implicitly so we probably do not need to change the use sites.

Yuhta replied:
The change should not be too large, you should only need to change the registry and existing deserialization implementations.  No need to change TypePtr because there is implicit conversion from unique_ptr<T> to shared_ptr<T>.  You can do the registry and other deserializer part in a separate PR.

duanmeng replied:
OK, got it. Add a new registry (`DeserializationRegistryForUniquePtr`) and update the Filters' create methods.
Will refactor the serialization framework using unique_ptr in a separate PR.

Yuhta commented on velox/type/Filter.h:
```diff
@@ -790,6 +915,42 @@ class BigintValuesUsingHashTable final : public Filter {
         values_(other.values_),
         sizeMask_(other.sizeMask_) {}
 
+  BigintValuesUsingHashTable(
```
We can use the existing ctor and set the additional fields in `create()` since it's in the same class.
In general we probably don't need extra ctor for `create()`.

Yuhta commented on velox/common/serialization/DeserializationRegistry.h:
```diff
@@ -38,6 +38,12 @@ using DeserializationWithContextRegistryType = Registry<
+    std::string,
+    std::unique_ptr<const ISerializable>(const folly::dynamic&)>;
+
+DeserializationRegistryUniquePtrType& DeserializationRegistryForUniquePtr();
```
nit: function name starts with lower case

duanmeng replied:
Fixed

Yuhta commented on velox/type/Filter.h:
```diff
@@ -55,18 +56,23 @@ enum class FilterKind {
 };
 
+class Filter;
+using FilterUniquePtr = std::unique_ptr<const Filter>;
```
No need to `const` qualify the `Filter` here

duanmeng replied:
Fixed

Yuhta commented on velox/common/serialization/Serializable.h:
```diff
@@ -177,6 +177,13 @@ class ISerializable {
           registryWithContext.Create(name, obj, context));
     }
 
+    const auto& registryOfUniquePtr =
```
Can we move this to a separate function like `std::unique_ptr<T> deserializeAsUniquePtr(const folly::dynamic& obj, void* context)`?

duanmeng replied:
OK, done.

duanmeng commented:
@Yuhta Hi Jimmy, I've resolved all the comments, could you please help to take a look?

duanmeng commented on velox/common/serialization/Serializable.h:
```diff
@@ -207,6 +213,24 @@ class ISerializable {
+  template <
+      class T,
+      typename = std::enable_if_t<std::is_base_of_v<ISerializable, T>>>
+  static auto deserializeAsUniquePtr(
```
Fixed

Yuhta commented on velox/type/Filter.h:
```diff
@@ -56,18 +57,23 @@ enum class FilterKind {
 };
 
+class Filter;
+using FilterUniquePtr = std::unique_ptr<Filter>;
```
@mbasmanova points out the convention is to use `FilterPtr`.  Can we use this name?

mbasmanova commented on velox/common/serialization/DeserializationRegistry.cpp:
```diff
@@ -17,15 +17,20 @@
 namespace facebook::velox {
 
-DeserializationRegistryType& DeserializationRegistryForSharedPtr() {
+DeserializationRegistryType& deserializationRegistryForSharedPtr() {
```
This change breaks Prestissimo and needs to be (1) in a separate PR; (2) implemented in backwards compatible way.

mbasmanova commented on velox/type/Filter.h:
```diff
@@ -638,6 +686,15 @@ class BigintRange final : public Filter {
+    auto ptr = dynamic_cast<const BigintRange*>(&other);
+    return Filter::testingEquals(other) && ptr != nullptr &&
+        (lower_ == ptr->lower_) && (upper_ == ptr->upper_) &&
+        (lower32_ == ptr->lower32_) && (upper32_ == ptr->upper32_) &&
```
Do we need to check lower32_ and others if lower_ and upper_ match?

mbasmanova commented on velox/type/Filter.h:
```diff
@@ -759,6 +830,13 @@ class HugeintRange final : public Filter {
+  bool testingEquals(const Filter& other) const final {
+    auto ptr = dynamic_cast<const HugeintRange*>(&other);
+    return Filter::testingEquals(other) && ptr != nullptr &&
+        (std::to_string(lower_) == std::to_string(ptr->lower_)) &&
```
Why do we have to convert lower_ and upper_ to strings to check for equality?

duanmeng replied:
Fixed

mbasmanova commented on velox/type/Filter.h:
```diff
@@ -830,6 +920,34 @@ class BigintValuesUsingHashTable final : public Filter {
         nullAllowed_ ? "with nulls" : "no nulls");
   }
 
+  bool testingEquals(const Filter& other) const final {
```
Can we move these implementations to .cpp file?

mbasmanova commented on velox/type/Filter.h:
```diff
@@ -822,6 +904,14 @@ class BigintValuesUsingHashTable final : public Filter {
+    return hashTable_;
+  }
+
+  void setHashTable(std::vector<int64_t> hashTable) {
```
Why do we need this method? Shouldn't we serialize values_ and then use the constructor?

duanmeng replied:
We could not reverse recover the same `BigintValuesUsingHashTable` using its current ctor `BigintValuesUsingHashTable(min, max, values, nullAllowed)`
for the order of the elements in **values** would affect the content of `BigintValuesUsingHashTable::hashTable_`, which is generated using **values** and linear probing.

During the construction of `BigintValuesUsingHashTable` it uses param **values** to

1.  use param **values** to initialize the `BigintValuesUsingHashTable::values_`
2.  use param **values** to initialize `BigintValuesUsingHashTable::hashTable_` using linear probing
3.  sort the `BigintValuesUsingHashTable::values_`

We could only use the sorted `BigintValuesUsingHashTable::values_` during serialization, so we may deserialize a different object if we use the current cotr `BigintValuesUsingHashTable(min, max, values, nullAllowed)`
So I add an extra ctor to to do a byte-by-byte copy using BigintValuesUsingHashTable(deterministic, nullAllowed, min, max, hashTable, containsEmptyMarker, values, sizeMask, values, nullAllowed);

For example, a `BigintValuesUsingHashTable` object named b1 constructed by the params `min = 13, max  = 9527, value = {7206, 7833, 6173, 3585, 7428, 3683, 7868, 1953, 6188}, nullAllowed = false`.
After construction, b1.values\_ = {1953, 3585, 3683, 6173, 6188, 7206, 7428, 7833, 7868}, <font color="red">b1.hashTable\_\[21] = 3585, b1.hashTable\_\[22] = 1953</font>.
However, when we use b1's member variables (`min_ = 13, max_  = 9527, value_ = {7206, 7833, 6173, 3585, 7428, 3683, 7868, 1953, 6188}, nullAllowed_ = false`)
to construct an `BigintValuesUsingHashTable` object named b2, <font color="red">b2.hashTable\_\[21] = 1953, b2.hashTable\_\[22] = 3585</font>, b1 and b2 would have different hashTable\_.

mbasmanova replied:
@duanmeng Thank you for explaining, but it seems that serializing sorted values and using that to re-create the object should produce logically equivalent filter, no? Should the testingEquals only compare values?

duanmeng replied:
Yes, they are logically equivalent. I will remove the setHashTable method and only compare values_ in the testingEquals method.

mbasmanova reviewed (commented):
@duanmeng Thank you for working on this. Some questions.

mbasmanova commented on velox/common/serialization/Serializable.h:
```diff
@@ -314,6 +320,25 @@ class ISerializable {
+      typename = std::enable_if_t<std::is_base_of_v<ISerializable, T>>>
+  static auto deserializeAsUniquePtr(
+      const folly::dynamic& obj,
+      void* context = nullptr) {
```
'context' appears unused here. Is this intentional? Shouldn't it be used?

mbasmanova replied:
Hmm... Shouldn't it be used? Since this is a generic API, I expect it to support deserializers that require additional context.

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
 }
 
+namespace {
+std::unordered_map<FilterKind, std::string> filterKinds() {
```
nit: perhaps, filterKindNames?

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+  };
+}
+
+const char* filterKinds(FilterKind kind) {
```
nit: perhaps, filterKinds -> filterKindName?

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+
+FilterPtr BoolValue::create(const folly::dynamic& obj) {
+  auto value = obj["value"].asBool();
+  auto nullAllowed = obj["nullAllowed"].asBool();
```
nit: I expect this code to repeat a lot; consider adding small helper function

```
auto nullAllowed = deserializeNullAllowed(obj);
```

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+  return otherBigintRange != nullptr && Filter::testingEquals(other) &&
+      (lower_ == otherBigintRange->lower_) &&
+      (upper_ == otherBigintRange->upper_) &&
+      (isSingleValue_ == otherBigintRange->isSingleValue_);
```
Is this check necessary? Assuming isSingleValue_ is derived from lower and upper_, this check will always pass, no?

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+  }
+  obj["values"] = values;
+
+  folly::dynamic hashTable = folly::dynamic::array;
```
Do we need to serialize the hash table?

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+    values.push_back(v.asInt());
+  }
+
+  std::vector<int64_t> hashTable;
```
This is not used, is it?

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+  auto max = obj["max"].asInt();
+
+  std::vector<int64_t> values;
+  auto valArr = obj["values"];
```
do not abbreviate: valArr -> valuesArray

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+}
+
+bool BigintValuesUsingHashTable::testingEquals(const Filter& other) const {
+  auto otherBigintValuesUsingHashTable =
```
nit: otherBigintValuesUsingHashTable -> otherBigintValues for readability

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+      Filter::testingEquals(other) &&
+      min_ == otherBigintValuesUsingHashTable->min_ &&
+      max_ == otherBigintValuesUsingHashTable->max_ &&
+      containsEmptyMarker_ ==
```
Is it necessary to check all these properties? Can we check only values?

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+    return false;
+  }
+
+  for (size_t i = 0; i < values_.size(); ++i) {
```
nit: perhaps, add a comment to explain that values_ are sorted, hence, can be compared pair-wise

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+  folly::dynamic values = folly::dynamic::array;
+  for (size_t i = 0; i < bitmask_.size(); ++i) {
+    if (bitmask_[i]) {
+      values.push_back(static_cast<int64_t>(i) + min_);
```
Do we need static_cast<int64_t> here?

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+  auto max = obj["max"].asInt();
+  auto nullAllowed = obj["nullAllowed"].asBool();
+
+  std::vector<int64_t> values;
```
This code seems to be repeating. Consider adding helper function: deserializeValues(obj)

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+}
+
+bool BigintValuesUsingBitmask::testingEquals(const Filter& other) const {
+  auto otherBigintValuesUsingBitmask =
```
otherBigintValuesUsingBitmask -> otherBigintValues

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+
+FilterPtr NegatedBigintValuesUsingHashTable::create(const folly::dynamic& obj) {
+  auto nullAllowed = obj["nullAllowed"].asBool();
+  auto deterministic = obj["deterministic"].asBool();
```
Is this used?

Do we need to serialize 'deterministic' flag?

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+  auto min = nonNegated->min();
+  auto max = nonNegated->max();
+  auto values = nonNegated->values();
+  auto hashTable = nonNegated->hashTable();
```
not used?

mbasmanova commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,645 @@ std::string Filter::toString() const {
+  auto obj = AbstractRange::serializeBase("FloatingPointRange");
+  obj["lower"] = lower_;
+  obj["upper"] = upper_;
+  obj["isDouble"] = false;
```
Do we need this?

duanmeng replied:
Got it, thanks.

mbasmanova commented on velox/type/Filter.h:
```diff
@@ -104,6 +110,15 @@ class Filter {
+   * Only used in test code.
+   * @return Whether an object is the same as itself.
+   */
+  virtual bool testingEquals(const Filter& other) const {
```
This method is a bit tricky. It is easy to forget to override it in a concrete Filter implementation and get incorrect results. I wonder if it would be safer to make it pure virtual and provide a different helper method to use in implementation of testingEquals.

duanmeng replied:
OK, make it pure virtual and helper method `Filter::testingBaseEquals`

mbasmanova reviewed (commented):
@duanmeng Thank you for iterating on this PR. Some follow-up comments.

mbasmanova reviewed (commented):
@duanmeng There are CI failures. Would you take a look? CC: @kgpai

Yuhta commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,625 @@ std::string Filter::toString() const {
+  auto obj = Filter::serializeBase("HugeintRange");
+  char lower[sizeof(int128_t)];
+  HugeInt::serialize(lower_, lower);
+  obj["lower"] = std::string(lower);
```
One more small fix
```suggestion
  obj["lower"] = std::string(lower, sizeof(int128_t)];
```

Yuhta commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,625 @@ std::string Filter::toString() const {
+  obj["lower"] = std::string(lower);
+  char upper[sizeof(int128_t)];
+  HugeInt::serialize(upper_, upper);
+  obj["upper"] = std::string(upper);
```
```suggestion
  obj["upper"] = std::string(upper, sizeof(int128_t));
```

duanmeng replied:
Fixed

kgpai commented:
> @duanmeng There are CI failures. Would you take a look? CC: @kgpai

Seems to be a Github actions issue (no logs and claims to be an internal error). cc @assignUser

duanmeng commented:
> > @duanmeng There are CI failures. Would you take a look? CC: @kgpai
> 
> Seems to be a Github actions issue (no logs and claims to be an internal error). cc @assignUser

@kgpai Yes, the failure jobs have no logs, I've rerun the *Ubuntu Benchmark* job in debug mode to get logs. Thanks for your help.

Yuhta commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,626 @@ std::string Filter::toString() const {
+  auto obj = Filter::serializeBase("HugeintRange");
+  auto size = sizeof(int128_t);
+  char lower[size];
+  HugeInt::serialize(lower_, lower);
```
A quick question, are we writing these serialization as JSON?  If yes, some of these strings are not printable

duanmeng replied:
No, HugeInt::serialize use memcpy to copy the int128 value's bytes into a char array.

```C++
static FOLLY_ALWAYS_INLINE int128_t deserialize(const char* serializedData) {
    int128_t value;
    memcpy(&value, serializedData, sizeof(int128_t));
    return value;
  }
```

duanmeng replied:
I've updated `HugeintRange` SerDe methods. Serialize the int128 value using to_string and deserialize it using the `HugeInt::stringToInt128` (a new added method in this pr ).
```C++
// Ser
obj["lower"] = std::to_string(lower_);
obj["upper"] = std::to_string(upper_);
// De
auto lower = HugeInt::stringToInt128(obj["lower"].asString());
auto upper = HugeInt::stringToInt128(obj["upper"].asString());
```
Now the output of toJson(hugeIntRangeObj) would be as follow,
`{"upper":"-555586388888891223462895","kind":"kHugeintRange","lower":"-111914766488860230291735206771233412079","nullAllowed":true,"name":"HugeintRange"}`

Yuhta replied:
As discussed in Slack, we can use 2 int64_t for lower and upper, or use one string for the whole int128

duanmeng commented:
No, HugeInt::serialize use memcpy to copy the int128 value's bytes into a char array.


获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Jimmy Lu ***@***.***>
发送时间: Friday, May 12, 2023 12:47:08 AM
收件人: facebookincubator/velox ***@***.***>
抄送: macduan ***@***.***>; Mention ***@***.***>
主题: Re: [facebookincubator/velox] Add serialization support for common::Filter (PR #4648)


@Yuhta commented on this pull request.

________________________________

In velox/type/Filter.cpp<https://github.com/facebookincubator/velox/pull/4648#discussion_r1191450212>:

> +  return std::make_unique<NegatedBigintRange>(lower, upper, nullAllowed);
+}
+
+bool NegatedBigintRange::testingEquals(const Filter& other) const {
+  auto otherNegatedBigintRange =
+      dynamic_cast<const NegatedBigintRange*>(&other);
+  return otherNegatedBigintRange != nullptr &&
+      Filter::testingBaseEquals(other) &&
+      nonNegated_->testingEquals(*(otherNegatedBigintRange->nonNegated_));
+}
+
+folly::dynamic HugeintRange::serialize() const {
+  auto obj = Filter::serializeBase("HugeintRange");
+  auto size = sizeof(int128_t);
+  char lower[size];
+  HugeInt::serialize(lower_, lower);


A quick question, are we writing these serialization as JSON? If yes, some of these strings are not printable

―
Reply to this email directly, view it on GitHub<https://github.com/facebookincubator/velox/pull/4648#pullrequestreview-1423039363>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABG65IFFIDONOBO7C2YWTP3XFUJYZANCNFSM6AAAAAAXCKZ6WE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Yuhta commented on velox/type/HugeInt.h:
```diff
@@ -52,6 +52,32 @@ class HugeInt {
     memcpy(serializedData, &value, sizeof(int128_t));
   }
+
+  static FOLLY_ALWAYS_INLINE int128_t stringToInt128(const std::string& str) {
```
nit: Rename it to `parse`?  Also put the implementation to cpp as there is no benefit inlining it.

Yuhta replied:
Also add some tests for it please

duanmeng replied:
Fixed

Yuhta commented on velox/type/HugeInt.h:
```diff
@@ -52,6 +52,32 @@ class HugeInt {
+    size_t idx = 0;
+
+    if (str.empty()) {
+      throw std::invalid_argument(
```
Use `VELOX_CHECK` etc.

duanmeng replied:
Fixed

Yuhta commented on velox/type/Filter.cpp:
```diff
@@ -98,6 +97,621 @@ std::string Filter::toString() const {
+
+  folly::dynamic values = folly::dynamic::array;
+  for (size_t i = 0; i < bitmask_.size(); ++i) {
+    auto j = bitmask_.size();
```
Remove unused value here

duanmeng replied:
Fixed

Yuhta commented on velox/type/HugeInt.cpp:
```diff
@@ -16,6 +16,32 @@
+        str[idx] >= '0' && str[idx] <= '9',
+        "Invalid character {} in the string.",
+        str[idx])
+    result = result * 10 + (str[idx] - '0');
```
This is failing under ASAN when parsing the mininum value: "signed integer overflow: 0x7ffffffffffffffffffffffffffffff8 + 8 cannot be represented in type '__int128'".

duanmeng replied:
Fixed, by adding early return logic in HugeInt::parse to avoid the potential *overflow* caused by *Type Widening* of int128_t making it more robust.
```C++
int128_t max = std::numeric_limits<int128_t>::max();
int128_t min = std::numeric_limits<int128_t>::min();
for (; idx < str.size(); ++idx) {
   ...
    int cur = str[idx] - '0';
    if ((result > max / 10) || (result == (max / 10) && cur > (max % 10))) {
      return negative ? min : max;
    }
    result = result * 10 + cur;
}
return negative ? -result : result;
```

Yuhta commented on velox/type/HugeInt.cpp:
```diff
@@ -16,6 +16,45 @@
+        std::isdigit(str[idx]), "Invalid character {} in the string.", str[idx])
+
+    int cur = str[idx] - '0';
+    // Only supports numbers in the range of [min, max], returns min or max if
```
We probably want to throw if the parsed value is out of range of int128_t

Yuhta commented on velox/type/HugeInt.cpp:
```diff
@@ -16,6 +16,65 @@
+
+  int128_t max = std::numeric_limits<int128_t>::max();
+  int128_t min = std::numeric_limits<int128_t>::min();
+  std::string errorMessage = fmt::format("{} is out of range of int128_t", str);
```
Do not allocate the string before we encounter the error
